File: backend/app/schemas/pds.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class PdsContextOut(BaseModel):
    env_id: str
    business_id: UUID
    created: bool
    source: str
    diagnostics: dict[str, Any] = Field(default_factory=dict)


class PdsProjectCreateRequest(BaseModel):
    env_id: str | None = None
    business_id: UUID | None = None
    program_id: UUID | None = None
    name: str = Field(min_length=2, max_length=240)
    project_code: str | None = Field(default=None, min_length=1, max_length=80)
    description: str | None = Field(default=None, max_length=4000)
    sector: str | None = Field(default=None, min_length=1, max_length=120)
    project_type: str | None = Field(default=None, min_length=1, max_length=120)
    stage: str = "planning"
    status: str = "active"
    project_manager: str | None = None
    start_date: date | None = None
    target_end_date: date | None = None
    approved_budget: Decimal = Field(default=Decimal("0"), ge=0)
    contingency_budget: Decimal = Field(default=Decimal("0"), ge=0)
    next_milestone_date: date | None = None
    currency_code: str = Field(default="USD", min_length=3, max_length=8)
    baseline_period: str | None = Field(default=None, min_length=5, max_length=20)
    baseline_lines: list["PdsBudgetLineIn"] = Field(default_factory=list)
    created_by: str | None = None


class PdsProjectOut(BaseModel):
    project_id: UUID
    env_id: UUID
    business_id: UUID
    program_id: UUID | None = None
    name: str
    project_code: str | None = None
    description: str | None = None
    sector: str | None = None
    project_type: str | None = None
    stage: str
    start_date: date | None = None
    target_end_date: date | None = None
    project_manager: str | None = None
    approved_budget: Decimal
    committed_amount: Decimal
    spent_amount: Decimal
    forecast_at_completion: Decimal
    contingency_budget: Decimal
    contingency_remaining: Decimal
    pending_change_order_amount: Decimal
    next_milestone_date: date | None = None
    risk_score: Decimal
    currency_code: str
    status: str
    created_at: datetime
    updated_at: datetime


class PdsProjectUpdateRequest(BaseModel):
    project_code: str | None = Field(default=None, min_length=1, max_length=80)
    name: str | None = Field(default=None, min_length=2, max_length=240)
    description: str | None = Field(default=None, max_length=4000)
    sector: str | None = Field(default=None, min_length=1, max_length=120)
    project_type: str | None = Field(default=None, min_length=1, max_length=120)
    stage: str | None = None
    status: str | None = None
    project_manager: str | None = None
    start_date: date | None = None
    target_end_date: date | None = None
    approved_budget: Decimal | None = Field(default=None, ge=0)
    contingency_budget: Decimal | None = Field(default=None, ge=0)
    next_milestone_date: date | None = None
    currency_code: str | None = Field(default=None, min_length=3, max_length=8)
    updated_by: str | None = None


class PdsBudgetLineIn(BaseModel):
    cost_code: str = Field(min_length=1, max_length=80)
    line_label: str = Field(min_length=1, max_length=240)
    approved_amount: Decimal = Field(default=Decimal("0"), ge=0)


class PdsBudgetBaselineRequest(BaseModel):
    period: str = Field(min_length=5, max_length=20)
    approved_budget: Decimal = Field(ge=0)
    lines: list[PdsBudgetLineIn] = Field(default_factory=list)
    created_by: str | None = None


class PdsBudgetRevisionRequest(BaseModel):
    period: str = Field(min_length=5, max_length=20)
    revision_ref: str = Field(min_length=1, max_length=120)
    amount_delta: Decimal
    reason: str | None = None
    status: str = "approved"
    created_by: str | None = None


class PdsContractCreateRequest(BaseModel):
    contract_number: str = Field(min_length=1, max_length=120)
    vendor_id: UUID | None = None
    vendor_name: str | None = None
    scope_description: str | None = None
    contract_value: Decimal = Field(default=Decimal("0"), ge=0)
    executed_date: date | None = None
    status: str = "active"
    created_by: str | None = None


class PdsCommitmentCreateRequest(BaseModel):
    period: str = Field(min_length=5, max_length=20)
    amount: Decimal
    contract_id: UUID | None = None
    created_by: str | None = None


class PdsChangeOrderCreateRequest(BaseModel):
    change_order_ref: str = Field(min_length=1, max_length=120)
    amount_impact: Decimal
    schedule_impact_days: int = 0
    approval_required: bool = True
    metadata_json: dict[str, Any] = Field(default_factory=dict)
    created_by: str | None = None


class PdsChangeOrderApproveRequest(BaseModel):
    approved_by: str | None = None


class PdsInvoiceCreateRequest(BaseModel):
    invoice_number: str = Field(min_length=1, max_length=120)
    amount: Decimal
    invoice_date: date | None = None
    status: str = "approved"
    created_by: str | None = None


class PdsPaymentCreateRequest(BaseModel):
    payment_ref: str = Field(min_length=1, max_length=120)
    amount: Decimal
    payment_date: date | None = None
    invoice_id: UUID | None = None
    status: str = "paid"
    created_by: str | None = None


class PdsForecastCreateRequest(BaseModel):
    period: str = Field(min_length=5, max_length=20)
    forecast_to_complete: Decimal
    eac: Decimal
    status: str = "published"
    created_by: str | None = None


class PdsMilestoneIn(BaseModel):
    milestone_name: str = Field(min_length=1, max_length=240)
    baseline_date: date | None = None
    current_date: date | None = None
    actual_date: date | None = None
    slip_reason: str | None = None
    is_critical: bool = False


class PdsScheduleBaselineRequest(BaseModel):
    period: str = Field(min_length=5, max_length=20)
    milestones: list[PdsMilestoneIn] = Field(default_factory=list)
    created_by: str | None = None


class PdsScheduleUpdateRequest(BaseModel):
    period: str = Field(min_length=5, max_length=20)
    milestones: list[PdsMilestoneIn] = Field(default_factory=list)
    created_by: str | None = None


class PdsRiskCreateRequest(BaseModel):
    risk_title: str = Field(min_length=1, max_length=240)
    probability: Decimal = Field(ge=0, le=1)
    impact_amount: Decimal = Field(default=Decimal("0"), ge=0)
    impact_days: int = Field(default=0, ge=0)
    mitigation_owner: str | None = None
    status: str = "open"
    created_by: str | None = None


class PdsSurveyResponseCreateRequest(BaseModel):
    survey_template_id: UUID | None = None
    vendor_name: str | None = None
    respondent_type: str = Field(min_length=1, max_length=120)
    score: Decimal | None = Field(default=None, ge=0, le=5)
    responses_json: dict[str, Any] = Field(default_factory=dict)
    created_by: str | None = None


class PdsSnapshotRunRequest(BaseModel):
    env_id: str | None = None
    business_id: UUID | None = None
    period: str = Field(min_length=5, max_length=20)
    project_id: UUID | None = None
    run_id: str | None = None
    created_by: str | None = None


class PdsSnapshotRunOut(BaseModel):
    run_id: str
    env_id: UUID
    business_id: UUID
    period: str
    project_id: UUID | None = None
    snapshot_hash: str
    portfolio_snapshot_id: UUID
    schedule_snapshot_id: UUID
    risk_snapshot_id: UUID
    vendor_snapshot_ids: list[UUID] = Field(default_factory=list)


class PdsReportPackRunRequest(BaseModel):
    env_id: str | None = None
    business_id: UUID | None = None
    period: str = Field(min_length=5, max_length=20)
    run_id: str | None = None
    created_by: str | None = None


class PdsReportPackRunOut(BaseModel):
    report_run_id: UUID
    env_id: UUID
    business_id: UUID
    period: str
    run_id: str
    snapshot_hash: str | None = None
    narrative_text: str | None = None
    artifact_refs_json: list[dict[str, Any]] = Field(default_factory=list)
    deterministic_deltas_json: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class PdsPortfolioOut(BaseModel):
    env_id: UUID
    business_id: UUID
    period: str
    approved_budget: Decimal
    committed: Decimal
    spent: Decimal
    eac: Decimal
    variance: Decimal
    contingency_remaining: Decimal
    open_change_order_count: int
    pending_approval_count: int
    top_risk_count: int


class PdsSiteReportCreateRequest(BaseModel):
    report_date: date
    summary: str | None = None
    blockers: str | None = None
    weather: str | None = None
    temperature_high: int | None = None
    temperature_low: int | None = None
    workers_on_site: int = Field(default=0, ge=0)
    work_performed: str | None = None
    delays: str | None = None
    safety_incidents: str | None = None
    created_by: str | None = None


class PdsRfiCreateRequest(BaseModel):
    rfi_number: str | None = Field(default=None, min_length=1, max_length=80)
    subject: str = Field(min_length=1, max_length=240)
    description: str | None = None
    assigned_to: str | None = None
    due_date: date | None = None
    priority: str = "normal"
    metadata_json: dict[str, Any] = Field(default_factory=dict)
    created_by: str | None = None


class PdsRfiUpdateRequest(BaseModel):
    subject: str | None = Field(default=None, min_length=1, max_length=240)
    description: str | None = None
    assigned_to: str | None = None
    due_date: date | None = None
    priority: str | None = None
    response_text: str | None = None
    status: str | None = None
    metadata_json: dict[str, Any] | None = None
    updated_by: str | None = None


class PdsSubmittalCreateRequest(BaseModel):
    vendor_id: UUID | None = None
    submittal_number: str | None = Field(default=None, min_length=1, max_length=80)
    description: str | None = None
    spec_section: str | None = None
    required_date: date | None = None
    submitted_date: date | None = None
    reviewed_date: date | None = None
    review_notes: str | None = None
    status: str = "pending"
    metadata_json: dict[str, Any] = Field(default_factory=dict)
    created_by: str | None = None


class PdsDocumentCreateRequest(BaseModel):
    rfi_id: UUID | None = None
    submittal_id: UUID | None = None
    title: str = Field(min_length=1, max_length=240)
    document_type: str = Field(default="general", min_length=1, max_length=120)
    version_label: str | None = Field(default=None, max_length=80)
    storage_key: str | None = None
    metadata_json: dict[str, Any] = Field(default_factory=dict)
    created_by: str | None = None


class PdsVendorCreateRequest(BaseModel):
    vendor_name: str = Field(min_length=1, max_length=240)
    trade: str | None = None
    license_number: str | None = None
    insurance_expiry: date | None = None
    contact_name: str | None = None
    contact_email: str | None = None
    status: str = "active"
    metadata_json: dict[str, Any] = Field(default_factory=dict)
    created_by: str | None = None


class PdsVendorUpdateRequest(BaseModel):
    vendor_name: str | None = Field(default=None, min_length=1, max_length=240)
    trade: str | None = None
    license_number: str | None = None
    insurance_expiry: date | None = None
    contact_name: str | None = None
    contact_email: str | None = None
    status: str | None = None
    metadata_json: dict[str, Any] | None = None
    updated_by: str | None = None


PdsProjectCreateRequest.model_rebuild()

File: repo-b/src/types/pds.ts
export interface PdsProjectListItem {
  project_id: string;
  env_id: string;
  business_id: string;
  program_id: string | null;
  name: string;
  project_code: string | null;
  description: string | null;
  sector: string | null;
  project_type: string | null;
  stage: string;
  status: string;
  project_manager: string | null;
  start_date: string | null;
  target_end_date: string | null;
  approved_budget: string;
  committed_amount: string;
  spent_amount: string;
  forecast_at_completion: string;
  contingency_budget: string;
  contingency_remaining: string;
  pending_change_order_amount: string;
  next_milestone_date: string | null;
  risk_score: string;
  currency_code: string;
  created_at: string;
  updated_at: string;
}

export type PdsProject = PdsProjectListItem;

export interface PdsBudgetLine {
  budget_line_id: string;
  budget_version_id: string;
  cost_code: string;
  line_label: string;
  approved_amount: string;
  committed_amount?: string;
  invoiced_amount?: string;
  paid_amount?: string;
}

export interface PdsBudgetSummary {
  approved_budget: string;
  committed_amount: string;
  spent_amount: string;
  forecast_at_completion: string;
  contingency_budget: string;
  contingency_remaining: string;
  pending_change_order_amount: string;
  variance: string;
  budget_used_ratio: string | number;
}

export interface PdsScheduleItem {
  milestone_id: string;
  milestone_name: string;
  baseline_date: string | null;
  current_date: string | null;
  actual_date: string | null;
  slip_reason: string | null;
  is_critical: boolean;
  created_at: string;
}

export interface PdsChangeOrder {
  change_order_id: string;
  project_id: string;
  change_order_ref: string;
  status: string;
  amount_impact: string;
  schedule_impact_days: number;
  approval_required: boolean;
  approved_at: string | null;
  created_at: string;
}

export interface PdsRfi {
  rfi_id: string;
  project_id: string;
  rfi_number: string;
  subject: string;
  description: string | null;
  assigned_to: string | null;
  due_date: string | null;
  priority: string;
  response_text: string | null;
  responded_at: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface PdsSubmittal {
  submittal_id: string;
  project_id: string;
  vendor_id: string | null;
  vendor_name?: string | null;
  submittal_number: string;
  description: string | null;
  spec_section: string | null;
  required_date: string | null;
  submitted_date: string | null;
  reviewed_date: string | null;
  review_notes: string | null;
  status: string;
  created_at: string;
}

export interface PdsSiteReport {
  site_report_id: string;
  project_id: string;
  report_date: string;
  summary: string | null;
  blockers: string | null;
  weather: string | null;
  temperature_high: number | null;
  temperature_low: number | null;
  workers_on_site: number;
  work_performed: string | null;
  delays: string | null;
  safety_incidents: string | null;
  created_at: string;
}

export interface PdsPunchItem {
  punch_item_id: string;
  project_id: string;
  title: string;
  status: string;
  assignee: string | null;
  due_date: string | null;
  created_at: string;
}

export interface PdsDocument {
  pds_document_id: string;
  project_id: string;
  rfi_id: string | null;
  submittal_id: string | null;
  title: string;
  document_type: string;
  version_label: string | null;
  storage_key: string | null;
  status: string;
  created_at: string;
}

export interface PdsVendor {
  vendor_id: string;
  env_id: string;
  business_id: string;
  vendor_name: string;
  trade: string | null;
  license_number: string | null;
  insurance_expiry: string | null;
  contact_name: string | null;
  contact_email: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface PdsProjectOverview {
  project: PdsProjectListItem;
  budget: PdsBudgetSummary;
  schedule: {
    schedule_health: string;
    total_slip_days: number;
    critical_flags: number;
    next_milestone_date: string | null;
    items: PdsScheduleItem[];
  };
  counts: {
    open_risks: number;
    high_risks: number;
    open_change_orders: number;
    pending_change_orders: number;
    open_rfis: number;
    overdue_rfis: number;
    site_report_count: number;
    team_size: number;
  };
  recent_activity: Array<{
    type: string;
    label: string | null;
    status: string | null;
    created_at: string | null;
  }>;
}

export interface PdsProjectDetail extends PdsProjectOverview {
  budget_detail?: {
    versions: Array<Record<string, unknown>>;
    lines: PdsBudgetLine[];
    revisions: Array<Record<string, unknown>>;
    commitments: Array<Record<string, unknown>>;
    invoices: Array<Record<string, unknown>>;
    payments: Array<Record<string, unknown>>;
    forecasts: Array<Record<string, unknown>>;
    change_orders: PdsChangeOrder[];
  };
}

export interface PdsPortfolioKpis {
  env_id: string;
  business_id: string;
  period: string;
  approved_budget: string;
  committed: string;
  spent: string;
  eac: string;
  variance: string;
  contingency_remaining: string;
  open_change_order_count: number;
  pending_approval_count: number;
  top_risk_count: number;
}

export interface PdsPortfolioDashboard {
  period: string;
  kpis: PdsPortfolioKpis & {
    active_project_count: number;
    project_count: number;
    projects_on_budget_pct: string | number;
    projects_on_schedule_pct: string | number;
    budget_used_ratio: string | number;
  };
  projects: Array<{
    project_id: string;
    name: string;
    project_code: string | null;
    sector: string | null;
    stage: string;
    status: string;
    project_manager: string | null;
    schedule_health: string;
    total_slip_days: number;
    budget_variance: string;
    budget_used_ratio: string | number;
    open_rfis: number;
    open_risks: number;
    pending_change_orders: number;
    next_milestone_date: string | null;
  }>;
  alerts: Array<{
    project_id: string;
    type: string;
    severity: string;
    message: string;
  }>;
  recent_activity: Array<{
    project_id: string;
    project_name: string;
    type: string;
    label: string | null;
    status: string | null;
    created_at: string | null;
  }>;
}

export interface PdsSnapshotRun {
  run_id: string;
  env_id: string;
  business_id: string;
  period: string;
  project_id: string | null;
  snapshot_hash: string;
  portfolio_snapshot_id: string;
  schedule_snapshot_id: string;
  risk_snapshot_id: string;
  vendor_snapshot_ids: string[];
}

export interface PdsReportPackRun {
  report_run_id: string;
  env_id: string;
  business_id: string;
  period: string;
  run_id: string;
  snapshot_hash: string | null;
  narrative_text: string | null;
  artifact_refs_json: Array<Record<string, unknown>>;
  deterministic_deltas_json: Record<string, unknown>;
  created_at: string;
}
